### PR TITLE
Firebase refactoring

### PR DESCRIPTION
**Pull request by iwi**
# Summary

Refactoring of firebase based on your comments and our discussion.

# Review

The last function keeps doing almost everything that was doing before, but the complexity of the contents has been extracted.
Some of the "simplification" doesn't feel necessary, but I'd like to discuss.

Also the previous `keep_url` function used to always return `False`. I've kept that outside and now the function returns nothing and the `is_yes` switch happens outside.

### CODE REVIEW

File: linkatos/firebase.py [hunk 1]
```diff
@@ -12,21 +12,36 @@ def initialise(api_key, project_name):
...
-    # creates token every time maybe worth doing it once every 30m as they
-    # expire every hour
-    auth = firebase.auth()
-    user = auth.sign_in_with_email_and_password(user, password)
-
-    db = firebase.database()
-    data = {
-        "url": url
-    }
+def authenticate_user(username, password, auth):
```
Comment (arnau) [inline linkatos/firebase.py]: I like `authenticate` better.

File: linkatos/firebase.py [hunk 2]
```diff
@@ -12,21 +12,36 @@ def initialise(api_key, project_name):
...
+    user = auth.sign_in_with_email_and_password(username, password)
```
Comment (arnau) [inline linkatos/firebase.py]: What about `auth.with_email`?

File: linkatos/firebase.py [hunk 3]
```diff
@@ -12,21 +12,36 @@ def initialise(api_key, project_name):
...
+
+def get_token(user):
+    return user['idToken']
+
+
+def store_url(url, db, token):
+    return db.child("users").push(to_data(url), token)
+
+
+def get_auth(firebase):
```
Comment (arnau) [inline linkatos/firebase.py]: I think this function adds no real value.

File: linkatos/firebase.py [hunk 4]
```diff
@@ -12,21 +12,36 @@ def initialise(api_key, project_name):
...
-    db = firebase.database()
-    data = {
-        "url": url
-    }
+def authenticate_user(username, password, auth):
+    user = auth.sign_in_with_email_and_password(username, password)
+    return user
+
+
+def get_db(firebase):
```
Comment (arnau) [inline linkatos/firebase.py]: I think this function adds no real value.

File: linkatos/firebase.py [hunk 5]
```diff
@@ -12,21 +12,36 @@ def initialise(api_key, project_name):
...
+    return {"url": url}
 
-    db.child("users").push(data, user['idToken'])
 
-    return False
+def connect_to_fb_and_store_url(url, username, password, firebase):
+    # the function should only be called if we need to store the url
+    auth = get_auth(firebase)
+    user = authenticate_user(username, password, auth)
+    token = get_token(user)
```
Comment (arnau) [inline linkatos/firebase.py]: I think, the unit you want is `token` or `get_token` if you like the verb. Or `user_token`.

At this level, I don't want to see the details for obtainint the token:

```py
credentials = {"username": username, "password": password} # I'd probably compose this outside this function as you want to treat them as a data unit.
token = user_token(firebase, credentials)
```

File: linkatos/firebase.py [hunk 6]
```diff
@@ -12,21 +12,36 @@ def initialise(api_key, project_name):
...
+    data = to_data(url)
```
Comment (arnau) [inline linkatos/firebase.py]: This is not used here right?

File: linkatos/firebase.py (hunk 1 above)
Comment (iwi) [inline linkatos/firebase.py]: fine

File: linkatos/firebase.py (hunk 2 above)
Comment (iwi) [inline linkatos/firebase.py]: It's a `pyrebase` function, I'm not convinced we should rename it.

File: linkatos/firebase.py (hunk 4 above)
Comment (iwi) [inline linkatos/firebase.py]: :+1:

File: linkatos/firebase.py (hunk 3 above)
Comment (iwi) [inline linkatos/firebase.py]: :+1:

File: linkatos/firebase.py (hunk 6 above)
Comment (iwi) [inline linkatos/firebase.py]: true... :S

File: linkatos/firebase.py (hunk 2 above)
Comment (arnau) [inline linkatos/firebase.py]: Ah, then it's out of our scope.

**Pull request closed without merging**